homeassistant/components/timer.py

```diff
@@ -0,0 +1,402 @@
+                        weeks=service.data.get(
+                            ATTR_WEEKS, DEFAULT_DURATION),
+                        days=service.data.get(
+                            ATTR_DAYS, DEFAULT_DURATION),
+                        hours=service.data.get(
+                            ATTR_HOURS, DEFAULT_DURATION),
+                        minutes=service.data.get(
+                            ATTR_MINUTES, DEFAULT_DURATION),
+                        seconds=service.data.get(
+                            ATTR_SECONDS, DEFAULT_DURATION),
+                    )
+                )
+        if tasks:
+            yield from asyncio.wait(tasks, loop=hass.loop)
+
+    descriptions = yield from hass.async_add_job(
+        load_yaml_config_file, os.path.join(
+            os.path.dirname(__file__), 'services.yaml')
+    )
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_START, async_handler_service,
+        descriptions[DOMAIN][SERVICE_START], SERVICE_SCHEMA_DURATION)
+    hass.services.async_register(
+        DOMAIN, SERVICE_PAUSE, async_handler_service,
+        descriptions[DOMAIN][SERVICE_PAUSE], SERVICE_SCHEMA)
+    hass.services.async_register(
+        DOMAIN, SERVICE_CANCEL, async_handler_service,
+        descriptions[DOMAIN][SERVICE_CANCEL], SERVICE_SCHEMA)
+    hass.services.async_register(
+        DOMAIN, SERVICE_FINISH, async_handler_service,
+        descriptions[DOMAIN][SERVICE_FINISH], SERVICE_SCHEMA)
+
+    yield from component.async_add_entities(entities)
+    return True
+
+
+class Timer(Entity):
+    """Representation of a timer."""
+
+    def __init__(self, hass, object_id, name, icon,
+                 weeks, days, hours, minutes, seconds):
+        """Initialize a timer."""
+        self.entity_id = ENTITY_ID_FORMAT.format(object_id)
+        self._name = name
+        self._state = STATUS_IDLE
+        self._duration = timedelta(weeks=weeks, days=days, hours=hours,
+                                   minutes=minutes, seconds=seconds)
+        self._remaining = self._duration
+        self._icon = icon
+        self._hass = hass
+        self._start = None
+        self._end = None
+        self._listener = None
+
+    @property
+    def should_poll(self):
+        """If entity should be polled."""
+        return True
+
+    @property
+    def name(self):
+        """Return name of the timer."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Return the icon to be used for this entity."""
+        return self._icon
+
+    @property
+    def state(self):
+        """Return the current value of the timer."""
+        return self._state
+
+    @property
+    def state_attributes(self):
+        """Return the state attributes."""
+        start = dt_util.as_local(self._start) if self._start else None
+        end = dt_util.as_local(self._end) if self._end else None
+        return {
+            ATTR_DURATION: self._duration.__str__(),
+            ATTR_START: start,
```

Comment:
Start doesn't make much sense. If you pause and then start a timer, start changes and it can no longer be calculated from the start?

I think that we just need duration and then in the frontend we know that `max(0, (last_updated + duration) - now)` is the time left on the timer.

Reply:
I agree, that start doesn't make sense for a paused timer. But I at least would like to keep the end so people can use templates to work with that when needed.

Reply:
Not sure what end is important? Templates are only getting re-rendered when entities change, not when time changes. I would expect a new automation trigger to be introduced that can respond to a timer with an optional offset.